Extract get_structure to abstract Mixin

pmrv opened the pull request:
A number of our classes define a `get_structure` with consistent semantics, but there was no one place for its definition.  I pulled it out into an ABC and added two other methods 

```python
number_of_structures
iter_structures
```

where the first gives the maximum "iteration_step" that can be passed to `get_structure` and the other just iterates over all structures.   Previously you would (and our code indeed does) check the size of the cells/positions array in the output of an atomistic job, but that's not so OOP.

On the implementation side, I've added extra methods

```python
_number_of_structures_impl
_get_structure_imp
```

that sub classes should override instead of the ones above, because that guarantees stable interfaces and correct docstrings.

Rationale: I want to have a common and guaranteed interface of obtaining structures from objects, for use in the structure/training container.  Both when adding structures to the container where something like

```python
container.include(job)
```

would be nice to have, but also when adding jobs to potential fitting jobs, I want structure/training containers to behave as the other atomistic jobs.

WIP: renaming all the previously defined `get_structures` to `_get_structure_impl` and deriving from `HasStructure`.

Christmas Wishlist: We also have a `structure` property, that I'd like to see on this class, but I thought touching `get_structure` is intrusive enough for now.

pmrv committed:
Add HasStructre mixin

pmrv committed:
Use HasStructure in AtomisticGenericJob

pmrv committed:
Flesh out HasStructure interface

get_number_of_structures() is now a property

get_structure and number_of_structures are now concrete methods, so that
they may perform additional argument validation and their docstrings are
preserved without bothering implementors

pmrv committed:
Use updated interface in AtomisticGenericJob

pmrv committed:
Use updated interface in AtomisticParallelMaster

pmrv committed:
Use HasStructure for Murnaghan classes

pmrv committed:
Use HasStructure interface in generic DFT job

pmrv committed:
Use HasStructure in VaspInteractive

liamhuber commented on pyiron_atomistics/atomistics/structure/has_structure.py:
```diff
@@ -0,0 +1,108 @@
+            IndexError: if not 0 < iteration_step < :property:`.number_of_structures`
+        """
+        num_structures = self.number_of_structures
+        if iteration_step < 0:
```
Edge case: Rather take a mod, as this will fail for numbers < -`number_of_structures`.

pmrv replied:
I only want to allow iteration_steps within [-N, N), so I do want this to fail for number < -N.  This is not reflected in the error message, so I'll touch that up, though.

liamhuber commented on pyiron_atomistics/atomistics/structure/has_structure.py:
```diff
@@ -0,0 +1,108 @@
+        return self._get_structure_impl(iteration_step=iteration_step, wrap_atoms=wrap_atoms)
+
+    @abstractmethod
+    def _get_structure_impl(self, iteration_step=-1, wrap_atoms=True):
```
I'm not very familiar with mixins... I assume `_impl` is short for "implicit"? Or? Is this standard practice for mixins? I could see rather just `_get_structure` instead (sim. for number of structures).

pmrv replied:
I put it as "implementation", but you're right dropping it, makes it shorter and it's still clearly private, so I'll do that.

liamhuber commented on pyiron_atomistics/atomistics/master/murnaghan.py:
```diff
@@ -590,7 +591,7 @@ def pouriertarantola(V, E0, B0, BP, V0):
 
 # ToDo: not all abstract methods implemented
-class Murnaghan(AtomisticParallelMaster):
+class Murnaghan(AtomisticParallelMaster, HasStructure):
```
I would really like that `AtomisticParallelMaster` has some `Atomistic` parent that already gives us the `HasStructure` inheritance. Doesn't have to happen in this PR though, and I haven't looked yet what the current tree is.

pmrv replied:
I'll check the tree if it is necessary.

liamhuber reviewed (commented):
I love this idea.

pmrv committed:
Rename private methods

pmrv committed:
Use HasStructure for SerialMaster

pmrv committed:
Use HasStructure for SQSJob

pmrv committed:
Use HasStructure for AtomisticExampleJob

pmrv committed:
Use HasStructure for Sphinx

pmrv committed:
Use HasStructure for GPAW

pmrv committed:
Document correct range for iteration_step

pmrv committed:
Don't derive from HasStructure again in Murnaghan

AtomisticParallelMaster already does that

pmrv committed:
Fix AtomisticExampleJob

There's at least one structure always returned, so return at least 1 in _number_of_strucures

pmrv committed:
Document how to document special behavior

pmrv committed:
Add test for murnaghan get_structure

pmrv commented:
The (newly added) murnaghan test case fails, because for the dummy jobs setup in the test suite the calculation does not converge (doesn't find a minimum), so I've checked manually whether it works. It does. I'll try to fiddle with the parameters, but in case it doesn't get better I suggest I'll take the test out again and we merge as is.

liamhuber commented on pyiron_atomistics/atomistics/job/atomistic.py:
```diff
@@ -9,6 +9,7 @@
 import warnings
 
 from pyiron_atomistics.atomistics.structure.atoms import Atoms
+from ..structure.has_structure import HasStructure
```
I don't have a specific complaint about the relative import, but it raises a yellow flag for me since I'm pretty sure we avoid it everywhere else.

liamhuber commented on pyiron_atomistics/atomistics/job/atomistic.py:
```diff
@@ -635,7 +625,7 @@ def get_structure(self, iteration_step=-1, wrap_atoms=True):
             except IndexError:
                 if wrap_atoms:
-                    raise IndexError('cell at step ', iteration_step, ' not found')
+                    raise IndexError('cell at step ', iteration_step, ' not found') from None
```
TIL about `from`ing exceptions 👍 
Not 100% sure how I feel about it, but I don't object at any rate.

pmrv replied:
It's nice I think when you're trying to keep the backtrace clean.  The original IndexError is an implementation detail, so it's nice to hide it.

liamhuber commented on pyiron_atomistics/atomistics/master/murnaghan.py:
```diff
@@ -11,6 +11,7 @@
 import warnings
 from pyiron_atomistics.atomistics.structure.atoms import Atoms, ase_to_pyiron
 from pyiron_atomistics.atomistics.master.parallel import AtomisticParallelMaster
+from ..structure.has_structure import HasStructure
```
`Murnaghan(AtomisticParallelMaster(ParallelMaster, AtomisticGenericJob(GenericJobCore, HasStructure)))` and `HasStructure` otherwise doesn't appear in `murnaghan.py`. Isn't this import redundant?

pmrv replied:
Yup, forgot to remove that after I dropped `HasStructure` from `Murnaghan`.

liamhuber commented on pyiron_atomistics/atomistics/master/parallel.py:
```diff
@@ -41,11 +41,11 @@ def structure(self, basis):
-            return self.structure
-        else:
-            raise ValueError("iteration_step should be either 0.")
+    def _get_structure(self, iteration_step=0, wrap_atoms=True):
```
I'm confused why `AtomisticParallelMaster` isn't an `ABC`, in which case these could (should?) be `abstractmethod`s. What am I missing?

pmrv replied:
In principle you could instantiate `AtomisticParallelMaster` directly with a custom generator, so it doesn't feel like an `ABC` to me.  But then again I just didn't want to touch too many things and just renamed the methods.

liamhuber commented on pyiron_atomistics/atomistics/structure/has_structure.py:
```diff
@@ -0,0 +1,116 @@
+
+    def get_structure(self, iteration_step=-1, wrap_atoms=True):
+        """
+        Gets the structure from a given iteration step of the simulation (MD/ionic relaxation). For static calculations
```
"(MD/ionic relaxation)" is too restrictive. Consider, e.g. Monte Carlo on chemistry. I would change this to "(e.g. MD, ionic relaxation, Monte Carlo, etc...)" or just remove it.

liamhuber replied:
(I lean weakly towards removal)

pmrv replied:
Agree, this also came up [here](https://github.com/pyiron/pyiron_contrib/pull/153).  Maybe it's best to deprecate `iteration_step` and rename it to something more generic together with the docstring.

liamhuber replied:
Yeah, I like `frame` personally

liamhuber commented on pyiron_atomistics/lammps/interactive.py:
```diff
@@ -51,11 +51,14 @@ def structure(self, structure):
 
-    def get_structure(self, iteration_step=-1, wrap_atoms=True):
-        return super(GenericInteractive, self).get_structure(
+    def _get_structure(self, iteration_step=-1, wrap_atoms=True):
```
I checked out the branch and it looks to me like `super()._get_structure` and `_number_of_structures` trace back to the methods defined on `AtomisticGenericJob`; is this correct? If so, why does the `super` call need an explicit parent path, and why are these methods needed at all since they're just naked supers calls?

pmrv replied:
Back then, I just renamed it blindly.  But I think you're right both inheritance paths go to `AtomisticGenericJob`, so the super calls should be unnecessary.

liamhuber commented on pyiron_atomistics/sphinx/interactive.py:
```diff
@@ -51,8 +51,8 @@ def structure(self, structure):
 
-    def get_structure(self, iteration_step=-1, wrap_atoms=True):
-        return GenericInteractive.get_structure(
+    def _get_structure(self, iteration_step=-1, wrap_atoms=True):
```
Yeah, in contrast I definitely see the need for the explicit super call here since the `_get_structure` method differs depending whether you wind up at `GenericDFTJob` or `AtomisticGenericJob`.

liamhuber reviewed (commented):
Looking good. There's a couple places I think I found some redundant code, so either you can clean things up a bit or I can learn something new 😉 It's all pretty minor stuff though, and I'm looking forward to having this in the codebase!

pmrv committed:
Clean up imports

pmrv committed:
Rename iteration_step to frame in get_structure

pmrv committed:
Rename iteration_step in _get_structure

pmrv committed:
Drop unnecessary super

Both LammpsBase and GenericInteractive have the same _get_structure from AtomisticGenericJob

pmrv committed:
Replace iteration_step also in docs/error
